third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::values::{DeserializationSeed, SerializationReadyValue, SizedID, Struct, Value};
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::value::{IdentifierMappingKind, MoveTypeLayout};
+use serde::{
+    de::{DeserializeSeed, Error as DeError},
+    ser::Error as SerError,
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::str::FromStr;
+
+pub trait CustomDeserialize {
+    fn custom_deserialize<'d, D: Deserializer<'d>>(
+        &self,
+        deserializer: D,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+    ) -> Result<Value, D::Error>;
+}
+
+pub trait CustomSerialize {
+    fn custom_serialize<S: Serializer>(
+        &self,
+        serializer: S,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+        sized_id: SizedID,
+    ) -> Result<S::Ok, S::Error>;
+}
+
+// FIXME: consolidate
+fn is_string_layout(layout: &MoveTypeLayout) -> bool {
+    use MoveTypeLayout::*;
+    if let Struct(move_struct) = layout {
+        if let [Vector(elem)] = move_struct.fields().iter().as_slice() {
+            if let U8 = elem.as_ref() {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+fn is_derived_string_struct_layout(layout: &MoveTypeLayout) -> bool {
+    use MoveTypeLayout::*;
+    if let Struct(move_struct) = layout {
+        if let [value_field, Vector(padding_elem)] = move_struct.fields().iter().as_slice() {
+            if is_string_layout(value_field) {
+                if let U8 = padding_elem.as_ref() {
+                    return true;
+                }
+            }
+        }
+    }
+    false
+}
+
+fn size_u32_as_uleb128(mut value: usize) -> usize {
+    let mut len = 1;
+    while value >= 0x80 {
+        // 7 (lowest) bits of data get written in a single byte.
```

Comment:
`7 (lowest) bits of data get written in a single byte.`

We are only right shifting `value` here? Not clear where we are writing the data in a single byte.